packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -0,0 +1,696 @@
+    @private
+  */
+  pushedData: function() {
+    this.send('pushedData');
+  },
+
+  flushChangedAttributes: function() {
+    this._inFlightAttributes = this._attributes;
+    this._attributes = Ember.create(null);
+  },
+
+  /**
+    @method adapterWillCommit
+    @private
+  */
+  adapterWillCommit: function() {
+    this.flushChangedAttributes();
+    this.send('willCommit');
+  },
+
+  /**
+    @method adapterDidDirty
+    @private
+  */
+  adapterDidDirty: function() {
+    this.send('becomeDirty');
+    this.updateRecordArraysLater();
+  },
+
+  /**
+    @method send
+    @private
+    @param {String} name
+    @param {Object} context
+  */
+  send: function(name, context) {
+    var currentState = get(this, 'currentState');
+
+    if (!currentState[name]) {
+      this._unhandledEvent(currentState, name, context);
+    }
+
+    return currentState[name](this, context);
+  },
+
+  notifyHasManyAdded: function(key, record, idx) {
+    if (this.record) {
+      this.record.notifyHasManyAdded(key, record, idx);
+    }
+  },
+
+  notifyHasManyRemoved: function(key, record, idx) {
+    if (this.record) {
+      this.record.notifyHasManyRemoved(key, record, idx);
+    }
+  },
+
+  notifyBelongsToChanged: function(key, record) {
+    if (this.record) {
+      this.record.notifyBelongsToChanged(key, record);
+    }
+  },
+
+  notifyPropertyChange: function(key) {
+    if (this.record) {
+      this.record.notifyPropertyChange(key);
+    }
+  },
+
+  rollback: function() {
+    var dirtyKeys = Ember.keys(this._attributes);
+
+    this._attributes = Ember.create(null);
+
+    if (get(this, 'isError')) {
+      this._inFlightAttributes = Ember.create(null);
+      this.didCleanError();
+    }
+
+    //Eventually rollback will always work for relationships
+    //For now we support it only out of deleted state, because we
+    //have an explicit way of knowing when the server acked the relationship change
+    if (this.isDeleted()) {
+      //TODO: Should probably move this to the state machine somehow
+      this.becameReady();
+      this.reconnectRelationships();
+    }
+
+    if (this.isNew()) {
+      this.clearRelationships();
+    }
+
+    if (this.isValid()) {
+      this._inFlightAttributes = Ember.create(null);
+    }
+
+    this.send('rolledBack');
+
+    this.record._notifyProperties(dirtyKeys);
+
+  },
+  /**
+    @method transitionTo
+    @private
+    @param {String} name
+  */
+  transitionTo: function(name) {
+    // POSSIBLE TODO: Remove this code and replace with
+    // always having direct reference to state objects
+
+    var pivotName = extractPivotName(name);
+    var currentState = get(this, 'currentState');
+    var state = currentState;
+
+    do {
+      if (state.exit) { state.exit(this); }
+      state = state.parentState;
+    } while (!state.hasOwnProperty(pivotName));
+
+    var path = splitOnDot(name);
+    var setups = [];
+    var enters = [];
+    var i, l;
+
+    for (i=0, l=path.length; i<l; i++) {
```

Comment:
might be interesting to split this method into several smaller ones, especially ones the isolate the tight loops.

Now that `this` is stable, we would benefit from betting jitting, assuming the function size isn't to large.

Clearly this should actually be the result of some testing, as right now it is just a hunch.